File: java/rocksjni/table_properties_collector_factory.cc
//  Copyright (c) Meta Platforms, Inc. and affiliates.
//
//  This source code is licensed under both the GPLv2 (found in the
//  COPYING file in the root directory) and Apache 2.0 License
//  (found in the LICENSE.Apache file in the root directory).

#include "java/rocksjni/table_properties_collector_factory.h"

#include "java/include/org_rocksdb_TablePropertiesCollectorFactory.h"
#include "java/rocksjni/cplusplus_to_java_convert.h"
#include "rocksdb/db.h"
#include "rocksdb/utilities/table_properties_collectors.h"

/*
 * Class:     org_rocksdb_TablePropertiesCollectorFactory
 * Method:    newCompactOnDeletionCollectorFactory
 * Signature: (JJD)J
 */
jlong Java_org_rocksdb_TablePropertiesCollectorFactory_newCompactOnDeletionCollectorFactory(
    JNIEnv *, jclass, jlong sliding_window_size, jlong deletion_trigger,
    jdouble deletion_ratio) {
  auto *wrapper = new TablePropertiesCollectorFactoriesJniWrapper();
  wrapper->table_properties_collector_factories =
      ROCKSDB_NAMESPACE::NewCompactOnDeletionCollectorFactory(
          sliding_window_size, deletion_trigger, deletion_ratio);
  return GET_CPLUSPLUS_POINTER(wrapper);
}

/*
 * Class:     org_rocksdb_TablePropertiesCollectorFactory
 * Method:    deleteCompactOnDeletionCollectorFactory
 * Signature: (J)J
 */
void Java_org_rocksdb_TablePropertiesCollectorFactory_deleteCompactOnDeletionCollectorFactory(
    JNIEnv *, jclass, jlong jhandle) {
  auto instance =
      reinterpret_cast<TablePropertiesCollectorFactoriesJniWrapper *>(jhandle);
  delete instance;
}

File: java/src/main/java/org/rocksdb/TablePropertiesCollectorFactory.java
//  Copyright (c) Meta Platforms, Inc. and affiliates.
//
//  This source code is licensed under both the GPLv2 (found in the
//  COPYING file in the root directory) and Apache 2.0 License
//  (found in the LICENSE.Apache file in the root directory).

package org.rocksdb;

public abstract class TablePropertiesCollectorFactory extends RocksObject {
  private TablePropertiesCollectorFactory(final long nativeHandle) {
    super(nativeHandle);
  }

  public static TablePropertiesCollectorFactory NewCompactOnDeletionCollectorFactory(
      final long sliding_window_size, final long deletion_trigger, final double deletion_ratio) {
    long handle =
        newCompactOnDeletionCollectorFactory(sliding_window_size, deletion_trigger, deletion_ratio);
    return new TablePropertiesCollectorFactory(handle) {
      @Override
      protected void disposeInternal(long handle) {
        TablePropertiesCollectorFactory.deleteCompactOnDeletionCollectorFactory(handle);
      }
    };
  }

  /**
   * Internal API. Do not use.
   * @param nativeHandle
   * @return
   */
  static TablePropertiesCollectorFactory newWrapper(final long nativeHandle) {
    return new TablePropertiesCollectorFactory(nativeHandle) {
      @Override
      protected void disposeInternal(long handle) {
        TablePropertiesCollectorFactory.deleteCompactOnDeletionCollectorFactory(handle);
      }
    };
  }

  private static native long newCompactOnDeletionCollectorFactory(
      final long slidingWindowSize, final long deletionTrigger, final double deletionRatio);

  private static native void deleteCompactOnDeletionCollectorFactory(final long handle);
}
